wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt

```diff
@@ -158,51 +160,72 @@ class Options(
 
   private fun canonicalizeValue(
     linker: Linker,
-    context: Field,
+    context: ProtoType,
+    isRepeated: Boolean,
     value: Any
   ): Any {
     when (value) {
       is OptionElement -> {
         val result = mutableMapOf<ProtoMember, Any>()
         val field = linker.dereference(context, value.name)
         if (field == null) {
-          linker.errors += "unable to resolve option ${value.name} on ${context.type}"
+          linker.errors += "unable to resolve option ${value.name} on $context"
         } else {
-          val protoMember = get(context.type!!, field)
-          result[protoMember] = canonicalizeValue(linker, field, value.value)
+          val protoMember = get(context, field)
+          result[protoMember] =
+            canonicalizeValue(linker, field.type!!, field.isRepeated, value.value)
         }
-        return coerceValueForField(context, result)
+        return coerceValueForField(context, result, isRepeated)
       }
 
       is Map<*, *> -> {
-        val result = mutableMapOf<ProtoMember, Any>()
-        for (entry in value) {
-          val name = entry.key as String
-          val field = linker.dereference(context, name)
-          if (field == null) {
-            linker.errors += "unable to resolve option $name on ${context.type}"
-          } else {
-            val protoMember = get(context.type!!, field)
-            result[protoMember] = canonicalizeValue(linker, field, entry.value!!)
+        if (context.isMap) {
+          // Map fields are defined with two optional entries: `key` and 'value'.
+          val mapFieldKeyAsString = value["key"]
+          val mapFieldValueAsString = value["value"]
+          val mapFieldKey =
+            if (mapFieldKeyAsString == null) null
+            else canonicalizeValue(
+              linker, context.keyType!!, isRepeated = false, mapFieldKeyAsString
+            )
+          val mapFieldValue =
+            if (mapFieldValueAsString == null) null
+            else canonicalizeValue(
+              linker, context.valueType!!, isRepeated = false, mapFieldValueAsString
+            )
+          return coerceValueForField(context, mapOf(mapFieldKey to mapFieldValue), isRepeated)
```

Comment:
The new piece of logic. The rest is refactoring `field` to `prototype` + `isRepeated`